### PR TITLE
Checkpoint M2: lookup table checkpoint

### PR DESCRIPTION
**Pull request by seiriosPlus**
A new feature for #11410, it does not complete currently.
project: https://github.com/PaddlePaddle/Paddle/issues/10868

### CODE REVIEW

File: paddle/fluid/operators/load_op.cc [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (jacquesqiao) [inline paddle/fluid/operators/load_op.cc]: 多了两行

Comment (seiriosPlus) [reply paddle/fluid/operators/load_op.cc]: done

File: paddle/fluid/operators/save_op.cc [hunk 2]
```diff
@@ -67,9 +69,27 @@ class SaveOp : public framework::OperatorBase {
...
+      SaveSelectedRows(scope, place, var);
+    } else {
+      PADDLE_ENFORCE(
+          false,
+          "SaveOp only support LoDTensor and SelectedRows, %s has wrong type",
+          iname);
+    }
+  }
+
+  void SaveLodTensor(const platform::Place &place,
```
Comment (gongweibao) [inline paddle/fluid/operators/save_op.cc:93]: Need unit test of it.

File: python/paddle/fluid/io.py [hunk 3]
```diff
@@ -804,7 +806,9 @@ def save_checkpoint(executor,
                     trainer_id,
                     trainer_args=None,
                     main_program=None,
-                    max_num_checkpoints=3):
+                    max_num_checkpoints=3,
+                    lookup_table=None,
+                    ps_endpoint_list=None):
```
Comment (gongweibao) [inline python/paddle/fluid/io.py:812]: Need comments about the arguments.

File: paddle/fluid/operators/detail/grpc_client.h [hunk 4]
```diff
@@ -165,6 +165,20 @@ class FetchBarrierProcessor : public BaseProcessor {
   std::unique_ptr<sendrecv::SendRecvService::Stub> stub_;
 };
 
+class CheckpointNotifyProcessor : public BaseProcessor {
```
Comment (gongweibao) [inline paddle/fluid/operators/detail/grpc_client.h]: Can we reuse SendProcessor?

**Review by gongweibao — CHANGES_REQUESTED** [review]
Need a unit test to test checkpoint notification.

**Comment by gongweibao** [comment]
Unit test can done in another pr since this pr is big now.

File: paddle/fluid/operators/save_op.cc (hunk 2 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/save_op.cc:93]: ```save_load_op_test``` is the unit test.

File: paddle/fluid/operators/detail/grpc_client.h (hunk 4 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/detail/grpc_client.h]: We can reuse ```SendProcessor``` at currently, but ```checkpoint``` may have some new features in the future, I am not sure we can use ```SendProcessor ``` all the time, so, I think we can use a new one.

File: python/paddle/fluid/io.py (hunk 3 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py:812]: done

File: paddle/fluid/operators/checkpoint_notify_op.cc [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+               const platform::Place& place) const override {
+    std::vector<std::string> epmap = Attr<std::vector<std::string>>("epmap");
+    std::string dir = Attr<std::string>("dir");
+    std::string lookup_table_name = Attr<std::string>("lookup_table");
+
+    distributed::RPCClient* rpc_client =
+        distributed::RPCClient::GetInstance<RPCCLIENT_T>();
+    for (size_t i = 0; i < epmap.size(); i++) {
+      VLOG(3) << "checkpoint notify sending " << dir << " to " << epmap[i];
+      auto serial_looku_table =
```
Comment (jacquesqiao) [inline paddle/fluid/operators/checkpoint_notify_op.cc]: serial_looku_table  ==> lookup_table_save_dir?

File: paddle/fluid/operators/checkpoint_notify_op.cc [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  }
+};
+
+class CheckpointNotifyOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() {
+    AddAttr<std::vector<std::string>>(
+        "epmap",
+        "(string vector, default  127.0.0.1:6164)"
+        "Server endpoints in the order of input variables for mapping")
```
Comment (jacquesqiao) [inline paddle/fluid/operators/checkpoint_notify_op.cc]: the comment should change. no input variables.

File: paddle/fluid/operators/checkpoint_notify_op.cc [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    std::vector<std::string> epmap = Attr<std::vector<std::string>>("epmap");
+    std::string dir = Attr<std::string>("dir");
+    std::string lookup_table_name = Attr<std::string>("lookup_table");
+
+    distributed::RPCClient* rpc_client =
+        distributed::RPCClient::GetInstance<RPCCLIENT_T>();
+    for (size_t i = 0; i < epmap.size(); i++) {
+      VLOG(3) << "checkpoint notify sending " << dir << " to " << epmap[i];
+      auto serial_looku_table =
+          string::Sprintf("%s/%s_%d", dir, lookup_table_name, i);
```
Comment (jacquesqiao) [inline paddle/fluid/operators/checkpoint_notify_op.cc]: add a vlog to log the dir name?  for checking.

File: paddle/fluid/operators/checkpoint_notify_op.cc [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        "epmap",
+        "(string vector, default  127.0.0.1:6164)"
+        "Server endpoints in the order of input variables for mapping")
+        .SetDefault({"127.0.0.1:6164"});
+    AddAttr<std::string>(
+        "dir", "(string, default '') indicate the folder checkpoint will use");
+    AddAttr<std::string>("lookup_table",
+                         "(string, default '') the lookup table name");
+    AddComment(R"DOC(
+Prefetch operator
```
Comment (jacquesqiao) [inline paddle/fluid/operators/checkpoint_notify_op.cc]: update the comment

File: paddle/fluid/operators/checkpoint_notify_op.cc (hunk 7 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/checkpoint_notify_op.cc]: This op should also support pserver checkpoint without distribute_lookup_table.

File: paddle/fluid/operators/save_op.cc [hunk 9]
```diff
@@ -112,17 +125,44 @@ class SaveOp : public framework::OperatorBase {
...
       framework::SerializeToStream(fout, tensor, dev_ctx);
     }
+    fout.close();
+  }
+
+  void SaveSelectedRows(const framework::Scope &scope,
+                        const platform::Place &place,
+                        framework::Variable *var) const {
+    auto *lt_var =
+        scope.FindVar("loopup_table_path")->GetMutable<std::string>();
```
Comment (jacquesqiao) [inline paddle/fluid/operators/save_op.cc]: loopup_table_path  => lookup_table_path

Comment (jacquesqiao) [reply paddle/fluid/operators/save_op.cc]: 这个名字有点特殊化了，可以弄个宏定义一下

File: paddle/fluid/operators/distributed/grpc_server.cc [hunk 10]
```diff
@@ -187,6 +187,48 @@ class RequestPrefetch final : public RequestBase {
...
 };
 
+class RequestCheckpointNotify final : public RequestBase {
+ public:
+  explicit RequestCheckpointNotify(GrpcService::AsyncService* service,
+                                   ::grpc::ServerCompletionQueue* cq,
+                                   RequestHandler* request_handler, int req_id)
+      : RequestBase(service, cq, request_handler, req_id), responder_(&ctx_) {
+    request_.reset(new VariableResponse(request_handler->scope(),
+                                        request_handler->dev_ctx(), true));
```
Comment (jacquesqiao) [inline paddle/fluid/operators/distributed/grpc_server.cc]:
```cpp
new VariableResponse(request_handler->scope(), request_handler->dev_ctx(), true)
```
true可以改成false，这里不需要使用local_scope

File: paddle/fluid/operators/distributed/grpc_server.cc [hunk 11]
```diff
@@ -187,6 +187,48 @@ class RequestPrefetch final : public RequestBase {
...
+
+  std::string GetReqName() override { return request_->Varname(); }
+
+  void Process() override {
+    auto scope = request_->GetMutableLocalScope();
+
+    std::string checkpoint_notify = request_->Varname();
+    std::string checkpoint_dir = request_->OutVarname();
+
+    framework::Variable* invar = nullptr;
```
Comment (jacquesqiao) [inline paddle/fluid/operators/distributed/grpc_server.cc]: 这两行可以删掉

File: paddle/fluid/operators/distributed/grpc_server.cc [hunk 12]
```diff
@@ -187,6 +187,48 @@ class RequestPrefetch final : public RequestBase {
...
+    framework::Variable* outvar = nullptr;
+
+    VLOG(4) << "RequestCheckpointNotify notify: " << checkpoint_notify
+            << ", dir: " << checkpoint_dir;
+
+    request_handler_->Handle(checkpoint_notify, scope, invar, &outvar,
```
Comment (jacquesqiao) [inline paddle/fluid/operators/distributed/grpc_server.cc]:
```cpp
request_handler_->Handle(checkpoint_notify, scope, nullptr, nullptr, checkpoint_dir);
```

File: paddle/fluid/operators/distributed/grpc_server.cc [hunk 13]
```diff
@@ -276,8 +320,8 @@ void AsyncGRPCServer::TryToRegisterNewOne(const std::string& rpc_name,
     return;
   }
 
-  VLOG(4) << "register send rpc_name:" << rpc_name
-          << ", handler:" << rpc_call_map_[kRequestSend];
+  LOG(INFO) << "TryToRegisterNewOne on RPC NAME: " << rpc_name
```
Comment (jacquesqiao) [inline paddle/fluid/operators/distributed/grpc_server.cc]: use VLOG(level)

Comment (jacquesqiao) [reply paddle/fluid/operators/distributed/grpc_server.cc]: 这个日志为啥需要修改？

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 14]
```diff
@@ -243,28 +247,40 @@ void ListenAndServOp::RunImpl(const framework::Scope &scope,
 
   PADDLE_ENFORCE(!rpc_service_);
   std::string endpoint = Attr<std::string>("endpoint");
+  int checkpoint_notify_id = Attr<int>(kCheckpointBlockId);
```
Comment (jacquesqiao) [inline paddle/fluid/operators/listen_and_serv_op.cc]: `checkpoint_notify_id` 这个名字不太好

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 15]
```diff
@@ -904,6 +907,26 @@ def _create_table_optimize_block(self, pserver_index, pserver_program,
...
         return table_opt_block
 
+    def _create_checkpoint_save_block(self, pserver_program, pre_block_idx):
+        """
+        create a new block to handle save checkpoint.
+        """
+        import os
+
+        pserver_program.global_block().create_var(
+            name="loopup_table_path",
```
Comment (jacquesqiao) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: `loopup_table_path ` 字符串有问题

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 16]
```diff
@@ -904,6 +907,26 @@ def _create_table_optimize_block(self, pserver_index, pserver_program,
...
+            persistable=True,
+            type=core.VarDesc.VarType.RAW)
+
+        checkpoint_save_block = pserver_program.create_block(pre_block_idx)
+        checkpoint_save_block.append_op(
+            type='save',
+            inputs={'X': [self.table_name]},
```
Comment (jacquesqiao) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: `table_name` 为啥是input，而且应该也不是var吧？

File: python/paddle/fluid/io.py [hunk 17]
```diff
@@ -1068,6 +1187,29 @@ def save_trainer_args(dirname, trainer_id, trainer_args):
 
 
 def load_trainer_args(checkpoint_dir, serial, trainer_id, trainer_args):
+    """
+    trainer will load some args from it's  independent directory,
```
Comment (jacquesqiao) [inline python/paddle/fluid/io.py]: `it's  independent` 多了一个空格 :)

File: python/paddle/fluid/io.py [hunk 18]
```diff
@@ -1009,6 +1031,55 @@ def load_persist_vars_without_grad(executor,
...
+    """
+    The parameter server will load lookup table's local file in 
+    selectedrows variable.
+
+    Args:
+        executor(Executor): The executor to run for loading persistable variables
+        dirname(str): The directory path
+        main_program(Program): Find the variable named table_name in main_program
+        pserver_id(int): the serial number in pserver_endpoints list
+        table_name(str): lookup table name
```
Comment (jacquesqiao) [inline python/paddle/fluid/io.py:1044]:
```python
Returns:
```
前加一行空格

File: paddle/fluid/operators/save_op.cc [hunk 19]
```diff
@@ -112,17 +125,44 @@ class SaveOp : public framework::OperatorBase {
...
 class SaveOpProtoMaker : public framework::OpProtoAndCheckerMaker {
  public:
   void Make() override {
-    AddInput("X", "(Tensor ) Input tensor to be saved");
+    AddInput("X", "(Tensor ) Input LoDTensor and SelectedRows to be saved");
     AddComment(R"DOC(
 Save operator
 
-This operator will serialize and write a tensor variable to file on disk.
+This operator will serialize and write a tensor/selected rows variable to file on disk.
```
Comment (jacquesqiao) [inline paddle/fluid/operators/save_op.cc]: LoDTensor / SelectedRows

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 16 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: self.table_name是var, 已修改attr

File: python/paddle/fluid/io.py (hunk 18 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py:1044]: done

File: paddle/fluid/operators/save_op.cc (hunk 19 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/save_op.cc]: done

File: python/paddle/fluid/io.py (hunk 17 above)
Comment (seiriosPlus) [reply python/paddle/fluid/io.py]: done

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 15 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: done

File: paddle/fluid/operators/listen_and_serv_op.cc (hunk 14 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/listen_and_serv_op.cc]: rename to checkpoint_notify_block_id

File: paddle/fluid/operators/distributed/grpc_server.cc (hunk 13 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/distributed/grpc_server.cc]: 这个日志之前是错误的， 不管接收到的```rpc_name```是什么，只打印``` handler:" << rpc_call_map_[kRequestSend]```

File: paddle/fluid/operators/distributed/grpc_server.cc (hunk 12 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/distributed/grpc_server.cc]: done

File: paddle/fluid/operators/distributed/grpc_server.cc (hunk 11 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/distributed/grpc_server.cc]: done

File: paddle/fluid/operators/distributed/grpc_server.cc (hunk 10 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/distributed/grpc_server.cc]: done

File: paddle/fluid/operators/save_op.cc (hunk 9 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/save_op.cc]: 用```constexpr ```做了优化， 放宏到```macros.h```中会引入不需要的rpc相关的头文件

File: paddle/fluid/operators/checkpoint_notify_op.cc (hunk 8 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/checkpoint_notify_op.cc]: done

File: paddle/fluid/operators/checkpoint_notify_op.cc (hunk 7 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/checkpoint_notify_op.cc]: 1. done.
2. 目前接口是用checkpoint_notify命名的，后期有需要可以优化。

File: paddle/fluid/operators/checkpoint_notify_op.cc (hunk 6 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/checkpoint_notify_op.cc]: done

File: paddle/fluid/operators/checkpoint_notify_op.cc (hunk 5 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/checkpoint_notify_op.cc]: done

File: paddle/fluid/operators/distributed/grpc_server.cc [hunk 20]
```diff
@@ -224,6 +266,8 @@ void AsyncGRPCServer::StartServer() {
     reqs.reserve(kRequestBufSize);
 
     for (int i = 0; i < kRequestBufSize; i++) {
+      LOG(INFO) << "TryToRegisterNewOne on RPC NAME: " << rpc_name
```
Comment (gongweibao) [inline paddle/fluid/operators/distributed/grpc_server.cc]: 这些日志太多了，用VLOG

File: paddle/fluid/operators/distributed/grpc_server.cc (hunk 13 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/grpc_server.cc]: req_id?

**Review by gongweibao — CHANGES_REQUESTED** [review]
日志打印有点小问题。

File: paddle/fluid/operators/distributed/grpc_server.cc (hunk 20 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/distributed/grpc_server.cc]: done

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 21]
```diff
@@ -910,6 +913,29 @@ def _create_table_optimize_block(self, pserver_index, pserver_program,
...
+            outputs={},
+            attrs={
+                'file_path':
```
Comment (jacquesqiao) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]:
```python
'file_path': ""
```
put `this 'file_path' do not be used in save lookup table variable` into comment

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 22]
```diff
@@ -242,23 +245,29 @@ void ListenAndServOp::RunImpl(const framework::Scope &scope,
 
   PADDLE_ENFORCE(!rpc_service_);
   std::string endpoint = Attr<std::string>("endpoint");
+  int checkpoint_notify_block_id = Attr<int>(kCheckpointBlockId);
 
   LOG(INFO) << "sync_mode:" << sync_mode << ", fan_in:" << fan_in
-            << ", end_point:" << endpoint;
+            << ", end_point:" << endpoint
+            << ", CheckpointNotify Id: " << checkpoint_notify_block_id;
```
Comment (jacquesqiao) [inline paddle/fluid/operators/listen_and_serv_op.cc]: "checkpoint_notify_block_id" => "checkpoint_block_id"
 "CheckpointNotify Id" => "checkpoint_block_id "

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 23]
```diff
@@ -267,6 +276,12 @@ void ListenAndServOp::RunImpl(const framework::Scope &scope,
   auto *program = optimize_blocks[0]->Program();
   framework::Executor executor(dev_place);
 
+  std::shared_ptr<framework::ExecutorPrepareContext> ckpt_pre_context = nullptr;
+  if (checkpoint_notify_block_id != -1) {
+    auto ctx = executor.Prepare(*program, checkpoint_notify_block_id);
+    ckpt_pre_context = std::move(ctx);
```
Comment (jacquesqiao) [inline paddle/fluid/operators/listen_and_serv_op.cc]: 看着有点奇怪，加点注释。

File: paddle/fluid/operators/save_op.cc [hunk 24]
```diff
@@ -22,11 +22,17 @@ limitations under the License. */
...
+#include "paddle/fluid/framework/selected_rows.h"
+#include "paddle/fluid/framework/variable.h"
 #include "paddle/fluid/platform/device_context.h"
 
 namespace paddle {
 namespace operators {
 
+// define LOOKUP_TABLE_PATH for checkpoint notify to save lookup table variables
+// to directory specified.
+constexpr char LOOKUP_TABLE_PATH[] = "lookup_table_path";
```
Comment (jacquesqiao) [inline paddle/fluid/operators/save_op.cc]: https://google.github.io/styleguide/cppguide.html#Constant_Names
`LOOKUP_TABLE_PATH ` => `kLookupTablePath`

Comment (seiriosPlus) [reply paddle/fluid/operators/save_op.cc]: done

File: paddle/fluid/operators/listen_and_serv_op.cc (hunk 23 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/listen_and_serv_op.cc]: done

File: paddle/fluid/operators/listen_and_serv_op.cc (hunk 22 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/listen_and_serv_op.cc]: done

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 21 above)
Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: done

**Review by jacquesqiao — DISMISSED** [review]
Generally LGTM, we can merge it now~ Thanks for the great work!

**Review by gongweibao — APPROVED** [review]

**Pull request merged**